From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Fri, 23 May 2025 12:01:42 +0900
Subject: [PATCH] Cache block state tags

Dreeam Note:
Init path type under "ReloadableServerResources#updateStaticRegistryTags" on the server starts is too early,
so that custom things injected by plugins, CraftEngine, for example, will not be applied to the cache.
So we need to init it again after the plugin loads (Double init, but it doesn't matter)

diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1350,6 +1350,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             // Paper end
             org.spigotmc.WatchdogThread.hasStarted = true; // Paper
             org.dreeam.leaf.config.LeafConfig.regSparkExtraConfig(); // Leaf - Leaf config
+            net.minecraft.world.level.block.Blocks.initPathType(); // Leaf - Cache block state tags
             // Paper start - Add onboarding message for initial server start
             if (io.papermc.paper.configuration.GlobalConfiguration.isFirstStart) {
                 LOGGER.info("*************************************************************************************");
diff --git a/net/minecraft/server/ReloadableServerResources.java b/net/minecraft/server/ReloadableServerResources.java
--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -124,5 +124,9 @@ public class ReloadableServerResources {
     public void updateComponentsAndStaticRegistryTags() {
         this.postponedTags.forEach(Registry.PendingTags::apply);
         this.newComponents.forEach(DataComponentInitializers.PendingComponents::apply);
+        // Leaf start - Cache block state tags
+        net.minecraft.world.level.block.Blocks.initPathType();
+        net.minecraft.world.level.block.Blocks.initTagFlag();
+        // Leaf end - Cache block state tags
     }
 }
diff --git a/net/minecraft/world/entity/vehicle/DismountHelper.java b/net/minecraft/world/entity/vehicle/DismountHelper.java
--- a/net/minecraft/world/entity/vehicle/DismountHelper.java
+++ b/net/minecraft/world/entity/vehicle/DismountHelper.java
@@ -55,7 +55,7 @@ public class DismountHelper {
 
     public static VoxelShape nonClimbableShape(final BlockGetter level, final BlockPos pos) {
         BlockState blockState = level.getBlockState(pos);
-        return !blockState.is(BlockTags.CLIMBABLE) && (!(blockState.getBlock() instanceof TrapDoorBlock) || !blockState.getValue(TrapDoorBlock.OPEN))
+        return ((blockState.tagFlag & (org.dreeam.leaf.util.BlockMasks.TRAP_DOOR_CL_OPEN | org.dreeam.leaf.util.BlockMasks.CLIMBABLE_TAG)) == 0) // Leaf - Cache block state tags
             ? blockState.getCollisionShape(level, pos)
             : Shapes.empty();
     }
diff --git a/net/minecraft/world/level/block/Blocks.java b/net/minecraft/world/level/block/Blocks.java
--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -5960,4 +5960,21 @@ public class Blocks {
             // Paper end - Perf: Optimize block state cache initialization
         });
     }
+
+    // Leaf start - Cache block state tags
+    public static void initTagFlag() {
+        for (Block block : BuiltInRegistries.BLOCK) {
+            for (BlockState blockState : block.getStateDefinition().getPossibleStates()) {
+                blockState.initTagFlag();
+            }
+        }
+    }
+    public static void initPathType() {
+        for (Block block : BuiltInRegistries.BLOCK) {
+            for (BlockState blockState : block.getStateDefinition().getPossibleStates()) {
+                blockState.initPathType();
+            }
+        }
+    }
+    // Leaf end - Cache block state tags
 }
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -491,6 +491,8 @@ public abstract class BlockBehaviour implements FeatureElement {
         private boolean emptyConstantCollisionShape;
         private VoxelShape constantCollisionShape;
         private boolean isSpecialCollidingBlock;
+        public byte pathType; // Leaf - Cache block state tags
+        public int tagFlag; // Leaf - Cache block state tags
 
         private static boolean isSpecialCollidingBlock(final net.minecraft.world.level.block.state.BlockBehaviour.BlockStateBase block) {
             // note: cannot reference Blocks here, as Blocks#<clinit> initialises the state caches in parallel
@@ -683,6 +685,15 @@ public abstract class BlockBehaviour implements FeatureElement {
             // Paper end - optimise collisions
         }
 
+        // Leaf start - Cache block state tags
+        public void initPathType() {
+            pathType = (byte) net.minecraft.world.level.pathfinder.WalkNodeEvaluator.getPathTypeFromState(this.asState()).ordinal();
+        }
+        public void initTagFlag() {
+            tagFlag = org.dreeam.leaf.util.BlockMasks.init(asState());
+        }
+        // Leaf end - Cache block state tags
+
         public Block getBlock() {
             return this.blockOwner; // Paper - optimise TypedInstance
         }
diff --git a/net/minecraft/world/level/pathfinder/PathTypeCache.java b/net/minecraft/world/level/pathfinder/PathTypeCache.java
--- a/net/minecraft/world/level/pathfinder/PathTypeCache.java
+++ b/net/minecraft/world/level/pathfinder/PathTypeCache.java
@@ -23,7 +23,7 @@ public class PathTypeCache {
     }
 
     private PathType compute(final BlockGetter level, final BlockPos pos, final int index, final long key) {
-        PathType pathType = WalkNodeEvaluator.getPathTypeFromState(level, pos);
+        PathType pathType = WalkNodeEvaluator.leaf$pathType(level, pos); // Leaf - Cache block state tags
         this.positions[index] = key;
         this.pathTypes[index] = pathType;
         return pathType;
diff --git a/net/minecraft/world/level/pathfinder/PathfindingContext.java b/net/minecraft/world/level/pathfinder/PathfindingContext.java
--- a/net/minecraft/world/level/pathfinder/PathfindingContext.java
+++ b/net/minecraft/world/level/pathfinder/PathfindingContext.java
@@ -26,7 +26,7 @@ public class PathfindingContext {
 
     public PathType getPathTypeFromState(final int x, final int y, final int z) {
         BlockPos pos = this.mutablePos.set(x, y, z);
-        return this.cache == null ? WalkNodeEvaluator.getPathTypeFromState(this.level, pos) : this.cache.getOrCompute(this.level, pos);
+        return this.cache == null ? WalkNodeEvaluator.leaf$pathType(this.level, pos) : this.cache.getOrCompute(this.level, pos); // Leaf - Cache block state tags
     }
 
     public BlockState getBlockState(final BlockPos pos) {
diff --git a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -513,6 +513,17 @@ public class WalkNodeEvaluator extends NodeEvaluator {
         return blockPathType;
     }
 
+    // Leaf start - Cache block state tags
+    private static final PathType[] PATH_TYPES = PathType.values();
+    public static PathType leaf$pathType(final BlockGetter level, final BlockPos pos) {
+        BlockState blockState = level.getBlockStateIfLoaded(pos);
+        if (blockState == null) {
+            return PathType.BLOCKED;
+        }
+        return PATH_TYPES[blockState.pathType];
+    }
+    // Leaf end - Cache block state tags
+
     protected static PathType getPathTypeFromState(final BlockGetter level, final BlockPos pos) {
         // Paper start - Do not load chunks during pathfinding
         BlockState blockState = level.getBlockStateIfLoaded(pos);
@@ -520,6 +531,11 @@ public class WalkNodeEvaluator extends NodeEvaluator {
             return PathType.BLOCKED;
         }
         // Paper end
+        // Leaf start - Cache block state tags
+        return getPathTypeFromState(blockState);
+    }
+    public static PathType getPathTypeFromState(final BlockState blockState) {
+        // Leaf end - Cache block state tags
         Block block = blockState.getBlock();
         if (blockState.isAir()) {
             return PathType.OPEN;
